piet-cairo/src/lib.rs

```diff
@@ -329,12 +330,28 @@ impl<'a> RenderContext for CairoRenderContext<'a> {
                 InterpolationMode::Bilinear => Filter::Bilinear,
             };
             surface_pattern.set_filter(filter);
-            let rect = rect.into();
-            rc.ctx.translate(rect.x0, rect.y0);
-            rc.ctx.scale(
-                rect.width() / (image.get_width() as f64),
-                rect.height() / (image.get_height() as f64),
+            let src_rect = {
+                let src_rect = src_rect.into();
+                if src_rect.area() == 0.0 {
```

Comment:
I'm not sure I love this. If "entire image" is desired, I think I'd prefer an Option. I'm also aware that this doesn't interact seamlessly with the `Into` magic.

Reply:
I think we could just make `src_rect` be `impl Into<Option<Rect>>`? It wouldn't work with all `Into<Rect>` types I don't think (I don't think there's a blanket `impl<T, S> Into<Option<S>> for T where T: Into<S>`?) but it might be easier than this.

Reply:
my bigger question, though, is whether or not this should be a new API, and `draw_image` should be the easy mode that just draws an image.

Reply:
> my bigger question, though, is whether or not this should be a new API, and `draw_image` should be the easy mode that just draws an image.

That would be another choice, would avoid breakage, and would parallel our approach to fill mode. Obviously a reasonable implementation strategy is to have two thin wrappers that do the `Into` and call into a common function that has `Option<Rect>`.

Also, to address a comment above, `Into<Option<Rect>>` doesn't work the same way as `Into<Rect>`. It will take a `Rect` and a `None` value, but not an arbitrary type that has `Into<Rect>`. This feels like a sharp edge to me.

Reply:
I agree that it's a drawback, but my hunch is that in the wild most of the actual `impl Into<Rect>` arguments will be actual `Rect`s, and so my subjective hunch is that the ergonomic advantage of accepting either `Rect` or `None` is better than having `Rect::ZERO` be an alias for `Rect::EVERYTHING`.

Reply:
I'd favor the two-thin-wrappers approach best, I think. Though I can imagine wanting a 'fast path' function that always draws the entire image, in addition to the sometimes-all (`Into<Option<Rect>>`) and always-done-the-math (`Into<Rect>`). I don't think the `Into<Rect>` papercut is *too* bad, but it does feel like the kind of edge case that would feel thoughtless to a new user. (One of those "Why would they do it that way?" moments)

Reply:
Okay, I think making this separate API is probably the best option.